File: ui-angular/src/app/entity/minimal-user.ts
import {UniqEntity} from '@app/entity/uniq-entity';

export class MinimalUser extends UniqEntity {

    public static create(): MinimalUser {
        return new MinimalUser(
            null,
            null,
            null,
            null,
            null,
            null
        );
    }

    public static copy(user?: MinimalUser): MinimalUser {
        return new MinimalUser(
            user ? user.id : null,
            user ? user.email : null,
            user ? user.firstName : null,
            user ? user.secondName : null,
            user ? user.thirdName : null,
            user ? user.birthday : null,
        );
    }

    private _namePresentation: string = null;

    constructor(
        id: number | null,
        public email: string | null,
        public firstName: string | null,
        public secondName: string | null,
        public thirdName: string | null,
        public birthday: Date | null,
    ) {
        super(id);
    }

    get namePresentation(): string {

        if (this._namePresentation == null) {
            this._namePresentation = this.buildNamePresentation();
        }

        return this._namePresentation;
    }

    private buildNamePresentation(): string {

        const names: string[] = [];

        if (this.firstName) {
            names.push(this.firstName);
        }

        if (this.secondName) {
            names.push(this.secondName);
        }

        if (this.thirdName) {
            names.push(this.thirdName);
        }

        return names.length > 0 ? names.join(' ') : this.email;
    }
}

File: ui_flutter/lib/src/jcrm/ui/flutter/entity/minimal_user.dart
import 'package:jcrm_ui_flutter/src/jcrm/ui/flutter/entity/uniq_entity.dart';
import 'package:jcrm_ui_flutter/src/jcrm/ui/flutter/resource/minimal_user_resource.dart';

class MinimalUser extends UniqEntity {

  static final MinimalUser none = MinimalUser(0, "", "", "", "", DateTime.now());

  String email;
  String firstName;
  String secondName;
  String thirdName;
  DateTime? birthday;

  MinimalUser(int id, this.email, this.firstName, this.secondName, this.thirdName, this.birthday) : super(id);

  MinimalUser.fromResource(MinimalUserResource resource)
      : email = resource.email,
        firstName = resource.firstName == null? "" : resource.firstName!,
        secondName = resource.secondName == null ? "" : resource.secondName!,
        thirdName = resource.thirdName == null ? "" : resource.thirdName!,
        birthday = resource.birthday == null ? null : DateTime.parse(resource.birthday!),
        super.fromResource(resource);

  List<String> names() {

    List<String> result = [];

    if (firstName.isNotEmpty) {
      result.add(firstName);
    }
    if (secondName.isNotEmpty) {
      result.add(firstName);
    }
    if (thirdName.isNotEmpty) {
      result.add(firstName);
    }

    return result;
  }
}
